python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py

```diff
@@ -188,259 +194,378 @@ def find_never_handled_root_asset_partitions(
-                        partition_key=candidate.partition_key,
-                    ).parent_partitions
-
-                    (
-                        updated_parent_asset_partitions,
-                        _,
-                    ) = self.instance_queryer.get_updated_and_missing_parent_asset_partitions(
-                        candidate, parent_asset_partitions
+            if unreconciled_ancestors:
+                conditions[
+                    ParentOutdatedAutoMaterializeCondition(
+                        waiting_on_asset_keys=frozenset(unreconciled_ancestors)
                     )
-                    updated_parents = {
-                        parent.asset_key for parent in updated_parent_asset_partitions
-                    }
+                ].update({candidate})
+        return conditions
 
-                    will_update_parents = set()
-                    for parent in parent_asset_partitions:
-                        if (
-                            AutoMaterializeDecisionType.from_conditions(
-                                conditions_by_asset_partition.get(parent)
-                            )
-                            == AutoMaterializeDecisionType.MATERIALIZE
-                        ):
-                            will_update_parents.add(parent.asset_key)
-
-                    if updated_parents or will_update_parents:
-                        conditions.add(
-                            ParentMaterializedAutoMaterializeCondition(
-                                updated_asset_keys=frozenset(updated_parents),
-                                will_update_asset_keys=frozenset(will_update_parents),
-                            )
-                        )
+    def get_max_materializations_exceeded_conditions_for_key(
+        self, asset_key: AssetKey, candidates: AbstractSet[AssetKeyPartitionKey], limit: int
+    ) -> Mapping[
+        MaxMaterializationsExceededAutoMaterializeCondition, AbstractSet[AssetKeyPartitionKey]
+    ]:
+        # the set of asset partitions which exceed the limit
+        rate_limited_asset_partitions = set(
+            sorted(candidates, key=lambda x: sort_key_for_asset_partition(self.asset_graph, x))[
+                limit:
+            ]
+        )
+        return {
+            MaxMaterializationsExceededAutoMaterializeCondition(): rate_limited_asset_partitions
+        }
 
-            # if the parents will not be resolved this tick
-            waiting_on_asset_keys = get_waiting_on_asset_keys(candidate)
-            if waiting_on_asset_keys:
-                conditions.add(
-                    ParentOutdatedAutoMaterializeCondition(
-                        waiting_on_asset_keys=waiting_on_asset_keys
-                    )
-                )
+    def get_auto_materialize_conditions_for_asset(
+        self,
+        asset_key: AssetKey,
+        will_materialize_mapping: Mapping[AssetKey, AbstractSet[AssetKeyPartitionKey]],
```

Comment:
a comment to clarify that this is the decisions for assets earlier in the toposort would help